crates/driver/src/tests/cases/protocol_fees.rs

```diff
@@ -187,3 +188,130 @@ async fn volume_protocol_fee_sell_order() {
 
     protocol_fee_test_case(test_case).await;
 }
+
+#[tokio::test]
+#[ignore]
+async fn price_improvement_fee_buy_out_market_order() {
+    let quote_sell_amount = 50000000000000000000u128;
+    let quote_buy_amount = 45000000000000000000u128;
+    let fee_policy = FeePolicy::PriceImprovement {
+        factor: 0.5,
+        // high enough so we don't get capped by volume fee
+        max_volume_factor: 1.0,
+        quote: PriceImprovementQuote {
+            sell_amount: quote_sell_amount.into(),
+            buy_amount: quote_buy_amount.into(),
+            fee: 1000000000000000000u128.into(),
+        },
+    };
+    let test_case = TestCase {
+        order_side: order::Side::Buy,
+        fee_policy,
+        order_sell_amount: 50000000000000000000u128.into(),
+        solver_fee: Some(10000000000000000000u128.into()),
+        quote_sell_amount: quote_sell_amount.into(),
+        quote_buy_amount: quote_buy_amount.into(),
+        executed: quote_buy_amount.into(),
+        // sell amount + quote fee * factor
+        executed_sell_amount: 50500000000000000000u128.into(),
+        // executed buy amount should match quote buy amount
+        executed_buy_amount: quote_buy_amount.into(),
+    };
+
+    protocol_fee_test_case(test_case).await;
+}
+
+#[tokio::test]
+#[ignore]
+async fn price_improvement_fee_sell_out_market_order() {
+    let quote_sell_amount = 50000000000000000000u128;
+    let quote_buy_amount = 45000000000000000000u128;
+    let fee_policy = FeePolicy::PriceImprovement {
+        factor: 0.5,
+        // high enough so we don't get capped by volume fee
+        max_volume_factor: 1.0,
+        quote: PriceImprovementQuote {
+            sell_amount: quote_sell_amount.into(),
+            buy_amount: quote_buy_amount.into(),
+            fee: 1000000000000000000u128.into(),
+        },
+    };
+    let order_sell_amount = 50000000000000000000u128;
+    let test_case = TestCase {
+        order_side: order::Side::Sell,
+        fee_policy,
+        order_sell_amount: order_sell_amount.into(),
+        solver_fee: Some(10000000000000000000u128.into()),
+        quote_sell_amount: quote_sell_amount.into(),
+        quote_buy_amount: quote_buy_amount.into(),
+        executed: 40000000000000000000u128.into(),
+        // executed sell amount should match order sell amount
+        executed_sell_amount: order_sell_amount.into(),
+        executed_buy_amount: 44558823529411764706u128.into(),
+    };
+
+    protocol_fee_test_case(test_case).await;
+}
```

Comment:
Partially fillable orders are missing.

Also can we add a test that shows shifting surplus into network fees doesn't change the protocol fee charged? E.g. the following two executions pay the same fee (assuming same underlying orders & quotes)

1. Sell 0.1 ETH receive 3500 DAI, pay 0.9 ETH in network fees
2. Sell 1 ETH receive 3500 DAI, pay 0 in network fees

Reply:
Partially fillable orders require some further refactoring. I will create a separate issue for this.